json-transport/src/lib.rs

```diff
@@ -149,36 +130,40 @@ where
     Item: for<'de> Deserialize<'de>,
     SinkItem: Serialize,
 {
-    Ok(new(TcpStream::connect(addr).compat().await?))
+    Ok(new(TcpStream::connect(addr).await?))
 }
 
 /// Listens on `addr`, wrapping accepted connections in JSON transports.
-pub fn listen<Item, SinkItem>(addr: &SocketAddr) -> io::Result<Incoming<Item, SinkItem>>
+pub async fn listen<A, Item, SinkItem>(addr: A) -> io::Result<Incoming<Item, SinkItem>>
 where
+    A: ToSocketAddrs,
     Item: for<'de> Deserialize<'de>,
     SinkItem: Serialize,
 {
-    let listener = TcpListener::bind(addr)?;
+    let listener = TcpListener::bind(addr).await?;
     let local_addr = listener.local_addr()?;
-    let incoming = listener.incoming().compat();
+    let incoming = Box::pin(listener.incoming());
     Ok(Incoming {
         incoming,
         local_addr,
         ghost: PhantomData,
     })
 }
 
+trait IncomingTrait: Stream<Item = io::Result<TcpStream>> + std::fmt::Debug + Send {}
+impl<T: Stream<Item = io::Result<TcpStream>> + std::fmt::Debug + Send> IncomingTrait for T {}
+
 /// A [`TcpListener`] that wraps connections in JSON transports.
+#[pin_project]
 #[derive(Debug)]
 pub struct Incoming<Item, SinkItem> {
-    incoming: Compat01As03<tokio_tcp::Incoming>,
+    #[pin]
+    incoming: Pin<Box<dyn IncomingTrait>>,
```

Comment:
Is dyn trait necessary? I'd prefer to not pay the overhead of dynamic dispatch if possible.

Reply:
`tokio_net::tcp::Incoming` is not exported any more, so I had to improvise.

Reply:
Wow -- I didn't even know Rust allowed you to return private items.